File: Buddy/buddy-flutter/lib/core/services/progression_service.dart
import 'dart:math' as math;

import 'package:shared_preferences/shared_preferences.dart';

import '../persistence/coin_wallet.dart';

/// Bono diario + sistema de XP/level del cuidador.
class ProgressionService {
  static const _xpKey        = 'buddy.caretaker.xp.v1';
  static const _lastLoginKey = 'buddy.lastLogin.v1';
  static const _streakKey    = 'buddy.streak.v1';

  // --- XP / Level -------------------------------------------------------

  static Future<int> xp() async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.getInt(_xpKey) ?? 0;
  }

  static Future<void> addXP(int amount) async {
    final prefs = await SharedPreferences.getInstance();
    final v = (prefs.getInt(_xpKey) ?? 0) + amount;
    await prefs.setInt(_xpKey, v < 0 ? 0 : v);
  }

  static Future<int> level() async {
    final x = await xp();
    return math.sqrt(x / 100.0).floor() + 1;
  }

  static Future<int> xpInCurrentLevel() async {
    final lvl = await level();
    final prevReq = math.pow(lvl - 1, 2).toInt() * 100;
    return (await xp()) - prevReq;
  }

  static Future<int> xpForNextLevel() async {
    final lvl = await level();
    final nextReq = math.pow(lvl, 2).toInt() * 100;
    final prevReq = math.pow(lvl - 1, 2).toInt() * 100;
    return nextReq - prevReq;
  }

  // --- Daily login -----------------------------------------------------

  static Future<int?> claimDailyBonusIfAvailable() async {
    final prefs = await SharedPreferences.getInstance();
    final now = DateTime.now();
    final lastIso = prefs.getString(_lastLoginKey);
    final last = lastIso != null ? DateTime.parse(lastIso) : null;
    var streak = prefs.getInt(_streakKey) ?? 0;

    if (last != null) {
      final sameDay = last.year == now.year && last.month == now.month && last.day == now.day;
      if (sameDay) return null;
      final yesterday = now.subtract(const Duration(days: 1));
      final wasYesterday = last.year == yesterday.year &&
          last.month == yesterday.month &&
          last.day == yesterday.day;
      streak = wasYesterday ? streak + 1 : 1;
    } else {
      streak = 1;
    }

    await prefs.setInt(_streakKey, streak);
    await prefs.setString(_lastLoginKey, now.toIso8601String());
    final bonus = math.min(50, 5 + streak * 2);
    await CoinWallet.add(bonus);
    return bonus;
  }

  static Future<int> streak() async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.getInt(_streakKey) ?? 0;
  }

  static Future<void> reset() async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.remove(_xpKey);
    await prefs.remove(_lastLoginKey);
    await prefs.remove(_streakKey);
  }
}

File: Buddy/buddy-ios/Sources/buddy/Core/Services/ProgressionService.swift
import Foundation

/// Daily login bonus + caretaker XP/level system.
struct ProgressionService {
    private static let xpKey = "buddy.caretaker.xp.v1"
    private static let lastLoginKey = "buddy.lastLogin.v1"
    private static let streakKey = "buddy.streak.v1"

    // MARK: - Caretaker XP / Level

    static var xp: Int {
        get { UserDefaults.standard.integer(forKey: xpKey) }
        set { UserDefaults.standard.set(newValue, forKey: xpKey) }
    }

    static func addXP(_ amount: Int) {
        xp = max(0, xp + amount)
    }

    static var level: Int {
        // Level n requires n*100 XP to reach
        Int((Double(xp) / 100.0).squareRoot()) + 1
    }

    static var xpInCurrentLevel: Int {
        let lvl = level
        let prevReq = Int(pow(Double(lvl - 1), 2)) * 100
        return xp - prevReq
    }

    static var xpForNextLevel: Int {
        let lvl = level
        let nextReq = Int(pow(Double(lvl), 2)) * 100
        let prevReq = Int(pow(Double(lvl - 1), 2)) * 100
        return nextReq - prevReq
    }

    // MARK: - Daily login

    static var lastLogin: Date? {
        get { UserDefaults.standard.object(forKey: lastLoginKey) as? Date }
        set { UserDefaults.standard.set(newValue, forKey: lastLoginKey) }
    }

    static var streak: Int {
        get { UserDefaults.standard.integer(forKey: streakKey) }
        set { UserDefaults.standard.set(newValue, forKey: streakKey) }
    }

    /// Returns the bonus amount if a new day has started, otherwise nil.
    @discardableResult
    static func claimDailyBonusIfAvailable() -> Int? {
        let cal = Calendar.current
        let now = Date()
        if let last = lastLogin {
            if cal.isDate(last, inSameDayAs: now) { return nil }
            // Continuous streak only if yesterday
            if let yesterday = cal.date(byAdding: .day, value: -1, to: now), cal.isDate(last, inSameDayAs: yesterday) {
                streak += 1
            } else {
                streak = 1
            }
        } else {
            streak = 1
        }
        lastLogin = now
        let bonus = min(50, 5 + streak * 2) // 5,7,9,11... cap 50
        CoinWallet.add(bonus)
        return bonus
    }

    static func reset() {
        UserDefaults.standard.removeObject(forKey: xpKey)
        UserDefaults.standard.removeObject(forKey: lastLoginKey)
        UserDefaults.standard.removeObject(forKey: streakKey)
    }
}
